node/bft-consensus/src/lib.rs

```diff
@@ -169,8 +174,104 @@ impl ExecutionState for MyExecutionState {
                 consensus_output.batches.len(),
                 consensus_output.sub_dag.leader.header.author,
             );
-            // TODO: get the output to snarkOS
-            // self.node.save_consensus(consensus_output).await;
+
+            /*
+            TODO: the following generally works, but there are some open points
+                1. currently, only blocks produced by the beacon are considered valid
+                2. can the Aleo mempool have a different set of transactions than ones accepted by the consensus?
+                3. we can't fail here, i.e. the checks by the TransactionValidator must be final
+                4. every validator can create a block, but should they? the downstream can create it on its own too
+
+            let consensus = self.consensus.clone();
+            let account = self.account.clone();
+            let next_block = tokio::task::spawn_blocking(move || {
+                // Collect all the transactions contained in the agreed upon batches.
+                let mut transactions = Vec::new();
+                for batch in consensus_output.batches {
+                    for batch in batch.1 {
+                        for transaction in batch.transactions {
+                            let bytes = BytesMut::from(&transaction[..]);
+                            // TransactionValidator ensures that the Message can be deserialized.
+                            let message = Message::<N>::deserialize(bytes).unwrap();
+
+                            let unconfirmed_transaction =
+                                if let Message::UnconfirmedTransaction(unconfirmed_transaction) = message {
+                                    unconfirmed_transaction
+                                } else {
+                                    // TransactionValidator ensures that the Message is an UnconfirmedTransaction.
+                                    unreachable!();
+                                };
+
+                            // TransactionValidator ensures that the Message can be deserialized.
+                            let transaction = unconfirmed_transaction.transaction.deserialize_blocking().unwrap();
+
+                            transactions.push(transaction);
+                        }
+                    }
+                }
+
+                // Attempt to add the batched transactions to the Aleo mempool.
```

Comment:
So the mempool now just collects the txs from the round until they are included in the next block? We might need to have stricter guarantees of when this is cleared or invalidated. E.g. what happens if the validator hasn't yet produced the block but already receives a batch for the next round (I'd assume this is possible in the partially synchronous model)? 🤔

Reply:
indeed, we'll probably need to be more clever here, but it's a good starting point IMO